src/hotspot/share/cds/archiveHeapWriter.cpp

```diff
@@ -186,47 +182,71 @@ void ArchiveHeapWriter::ensure_buffer_space(size_t min_bytes) {
   _buffer->at_grow(to_array_index(min_bytes));
 }
 
-void ArchiveHeapWriter::copy_roots_to_buffer(GrowableArrayCHeap<oop, mtClassShared>* roots) {
-  Klass* k = Universe::objectArrayKlass(); // already relocated to point to archived klass
-  int length = roots->length();
-  _heap_roots_word_size = objArrayOopDesc::object_size(length);
-  size_t byte_size = _heap_roots_word_size * HeapWordSize;
-  if (byte_size >= MIN_GC_REGION_ALIGNMENT) {
-    log_error(cds, heap)("roots array is too large. Please reduce the number of classes");
-    vm_exit(1);
-  }
+objArrayOop ArchiveHeapWriter::manifest_root_segment(size_t offset, int element_count) {
```

Comment:
The meaning of `manifest` is not very clear to me, and I can't find `manifest` used in this sense in any HotSpot header files. Maybe use `allocate_root_segment` instead?

Reply:
Yeah, I wanted to avoid `allocate` to avoid the implication we are actually entering a proper allocation path. We are "just" reinterpreting the piece of buffer code as the oop here. But I am not insisting, so I renamed back to `allocate_root_segment`.